lib/log_ratelimit.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+
+#include <rte_cycles.h>
+#include <rte_lcore.h>
+#include <rte_memory.h>
+
+#include "gatekeeper_main.h"
+#include "gatekeeper_log_ratelimit.h"
+
+struct log_ratelimit_state {
+	uint64_t interval_cycles;
+	uint32_t burst;
+	uint32_t printed;
+	uint64_t begin;
+} __rte_cache_aligned;
+
+static struct log_ratelimit_state log_ratelimit_states[RTE_MAX_LCORE];
+
+void
+log_ratelimit_state_init(unsigned lcore_id, uint32_t interval, uint32_t burst)
+{
+	struct log_ratelimit_state *lrs;
+
+	RTE_VERIFY(lcore_id < RTE_MAX_LCORE);
+
+	lrs = &log_ratelimit_states[lcore_id];
+	lrs->interval_cycles = interval * cycles_per_ms;
+	lrs->burst = burst;
+	lrs->printed = 0;
+	lrs->begin = rte_rdtsc();
+}
+
+/*
+ * Rate limiting log entries.
+ *
+ * Returns:
+ * - true means go ahead and do it.
+ * - false means callbacks will be suppressed.
+ */
+static bool
+log_ratelimit_allow(struct log_ratelimit_state *lrs)
+{
+	uint64_t now;
+
+	if (lrs->interval_cycles == 0)
+		return true;
+
+	now = rte_rdtsc();
+
+	if (lrs->begin + lrs->interval_cycles < now) {
+		lrs->begin = now;
+		lrs->printed = 0;
```

Comment:
Let's rename `begin` with `end`, and do the following:
1. the test above becomes `lrs->end < now`; and
2. replace the body of this `if` statement with `log_ratelimit_reset(lrs, now);`. We will no longer need the braces for the body of this `if` statement;
3. Define `log_ratelimit_reset()` before `log_ratelimit_state_init()` as follows:
```
static inline void
log_ratelimit_reset(struct log_ratelimit_state *lrs, uint64_t now)
{
	lrs->printed = 0;
	lrs->end = now + lrs->interval_cycles;
}
```